openassessment/xblock/studio_mixin.py

```diff
@@ -45,12 +45,69 @@ class StudioMixin(object):
     # modules that are enabled, we need to keep track of the order
     # that the user left assessments in the editor, including
     # the ones that were disabled.  This allows us to keep the order
-    # that the user specified.
-    editor_assessments_order = List(
+    # that the user specified. We use a property in order to add validation
+    # logic, and update old problems with new validation requirements on load
+    _editor_assessments_order = List(
         default=DEFAULT_EDITOR_ASSESSMENTS_ORDER,
         scope=Scope.content,
         help="The order to display assessments in the editor."
     )
+    _editor_assessments_order_validated = False
+
+    def _get_editor_assessments_order(self):
+        if not self._editor_assessments_order_validated:
+            self._set_editor_assessments_order(self._editor_assessments_order)
+        return self._editor_assessments_order
+
+    def _set_editor_assessments_order(self, new_order):
+        new_order = self._validate_assessments_order(new_order)
+        self._editor_assessments_order = new_order
+
+    def _validate_assessments_order(self, order_to_validate):
+        # Backwards compatibility - may be loading a module without staff-assessment
+        if u'staff-assessment' not in order_to_validate:
+            order_to_validate.append(u'staff-assessment')
+            logger.info('Backwards compatibility: editor_assessments_order now contains staff-assessment')
+
+        # Backwards compatibility:
+        # If the problem already contains example-based assessment
+        # then allow the editor to display example-based assessments.
+        enabled_assessments = [asmnt['name'] for asmnt in self.valid_assessments]
+        if u'example-based-assessment' in enabled_assessments and u'example-based-assessment' not in order_to_validate:
+            order_to_validate.insert(0, u'example-based-assessment')
+            logger.info('Backwards compatibility: editor_assessments_order now contains example-based-assessment')
+
+        # Forwards compatibility:
+        # Include any additional assessments that may have been added since the problem was created.
+        new_assessments = list(set(DEFAULT_EDITOR_ASSESSMENTS_ORDER) - set(order_to_validate))
+        if len(new_assessments) > 0:
+            order_to_validate = new_assessments + order_to_validate
+            logger.info('Forwards compatibility: editor_assessments_order now includes {}'.format(new_assessments))
+
+        # Consistency
+        # If the editor assessments order doesn't match the problem order,
```

Comment:
Won't this prevent you from being able to change the order of the steps in Studio?